### PR TITLE
Persistent user config

### PR DESCRIPTION
**Pull request by harshilsharma63**
#### Summary
Added API for updating user config to allow for server-side, persisted user-config. This is meant to be a cross-browser alternate for browser local storage.

#### Ticket Link
https://community-daily.mattermost.com/boards/workspace/qgsck6cts3fwpqwyjiupjm5cde/b4ctmm917q3rzdfi451qtuxjzcw/vpb5hydjzqtr7txbxeqpjxyfz4o/cn5nib48arpyu5rog6mx5ba3feo

### CODE REVIEW

File: webapp/src/pages/welcome/welcomePage.test.tsx [hunk 1]
```diff
@@ -8,19 +8,28 @@ import {createMemoryHistory} from 'history'
...
+
 const w = (window as any)
 const oldBaseURL = w.baseURL
 
+global.fetch = FetchMock.fn
+
 beforeEach(() => {
-    UserSettings.welcomePageViewed = null
+    FetchMock.fn.mockReset()
+    FetchMock.fn.mockReturnValue(FetchMock.jsonResponse(JSON.stringify({})))
```
Comment (kamre) [inline webapp/src/pages/welcome/welcomePage.test.tsx:32]: May be mocking `mutator` like in many other unit tests will be a little bit better.

Comment (harshilsharma63) [reply webapp/src/pages/welcome/welcomePage.test.tsx:32]: @kamre I'm having a hard time finding where we are mocking `mutator`. Can you share one such example?

Comment (harshilsharma63) [reply webapp/src/pages/welcome/welcomePage.test.tsx:32]: Ah, got it! Thanks for pointing that out. That definitely looks much better than emulating fetch. I'll make the change 👍

Comment (harshilsharma63) [reply webapp/src/pages/welcome/welcomePage.test.tsx]: Done.

File: server/api/api.go [hunk 2]
```diff
@@ -460,6 +461,82 @@ func (a *API) handlePostBlocks(w http.ResponseWriter, r *http.Request) {
...
+
+	vars := mux.Vars(r)
+	userID := vars["userID"]
+
+	ctx := r.Context()
+	session := ctx.Value(sessionContextKey).(*model.Session)
+
+	auditRec := a.makeAuditRecord(r, "updateUserConfig", audit.Fail)
+	defer a.audit.LogRecord(audit.LevelRead, auditRec)
+	auditRec.AddMeta("performedBy", userID)
```
Comment (wiggin77) [inline server/api/api.go:515]: We need to keep the fields names as consistent as possible since customers query by them.  These would be `user_id` and `session_id`.

File: server/server/server.go [hunk 3]
```diff
@@ -233,13 +232,6 @@ func NewStore(config *config.Configuration, logger *mlog.Logger) (store.Store, e
 	if err != nil {
 		return nil, err
 	}
-	if config.AuthMode == MattermostAuthMod {
```
Comment (wiggin77) [inline server/server/server.go:236]: Is this unrelated?

File: server/swagger/swagger.yml [hunk 4]
```diff
@@ -109,6 +109,23 @@ definitions:
         x-go-name: UpdatedFields
     type: object
     x-go-package: github.com/mattermost/focalboard/server/model
+  UserPropPatch:
+    description: UserConfigPatch is a patch for use config
```
Comment (wiggin77) [inline server/swagger/swagger.yml:113]:
```suggestion
    description: UserConfigPatch is a patch for user config
```

File: webapp/i18n/en.json [hunk 5]
```diff
@@ -240,7 +240,8 @@
   "ViewTitle.show-description": "show description",
   "ViewTitle.untitled-board": "Untitled board",
   "WelcomePage.Description": "Boards is a project management tool that helps define, organize, track and manage work across teams, using a familiar kanban board view",
-  "WelcomePage.Explore.Button": "Explore",
+  "WelcomePage.Explore.Button": "Take a tour",
```
Comment (wiggin77) [inline webapp/i18n/en.json:243]: Unrelated?

File: webapp/src/app.tsx [hunk 6]
```diff
@@ -128,7 +128,7 @@ const App = React.memo((): JSX.Element => {
     }
 
     const continueToWelcomeScreen = () => {
-        return Utils.isFocalboardPlugin() && loggedIn === true && !UserSettings.welcomePageViewed
+        return Utils.isFocalboardPlugin() && loggedIn === true && !me?.props[UserSettingKey.WelcomePageViewed]
```
Comment (wiggin77) [inline webapp/src/app.tsx:131]: Also unrelated?  Is this PR based on another PR that is unmerged?

**Review by wiggin77 — CHANGES_REQUESTED** [review]
It looks like there are some welcome page changes mixed in here.  I suspect this PR was based on another that is not yet merged .

File: server/api/api.go (hunk 2 above)
Comment (harshilsharma63) [reply server/api/api.go:515]: `user_id` and `session_id` are already being included by `LogRecord()` method as general audit metadata. The user ID and session user ID are logged here in the context of this specific operation. I know user ID is already included by `LogRecord` but I still wanna keep it here for context.

Would renaming here as `performed_by` and `performed_on` resolve your concern?

File: server/server/server.go (hunk 3 above)
Comment (harshilsharma63) [reply server/server/server.go:236]: This is intentional. `NewStore` isn't used in plugin mode. This previous code was doing nothing.

File: server/swagger/swagger.yml (hunk 4 above)
Comment (harshilsharma63) [reply server/swagger/swagger.yml:113]: Done.

File: webapp/i18n/en.json (hunk 5 above)
Comment (harshilsharma63) [reply webapp/i18n/en.json:243]: Nope, totally related. We're changing the text of buttons that trigger the change in user config.

File: webapp/src/app.tsx (hunk 6 above)
Comment (harshilsharma63) [reply webapp/src/app.tsx:131]: Nope, totally related. We're moving `WelcomePageViewed` from being managed in browser localstorage to server side user props.

**Comment by harshilsharma63** [comment]
> It looks like there are some welcome page changes mixed in here. I suspect this PR was based on another that is not yet merged .

Nope, nothing is mixed in. Everything is related to one single goal - adding server-side API for persisting user config and migrating one config (`WelcomePageViewed`) to use this API.

File: server/api/api.go (hunk 2 above)
Comment (wiggin77) [reply server/api/api.go]: When customers query these audit calls they know `user_id` and `session_id` answer the "who did this".  Adding duplicate fields with the same info doesn't break anything, but potentially confuses the customer.  They wonder, "can `performed_by` ever to be different than `user_id`?  Do I need to special case this API when querying?"

**Comment by wiggin77** [comment]
> > It looks like there are some welcome page changes mixed in here. I suspect this PR was based on another that is not yet merged .
> 
> Nope, nothing is mixed in. Everything is related to one single goal - adding server-side API for persisting user config and migrating one config (`WelcomePageViewed`) to use this API.

Ok, understood.  The PR description only included an API being added.  It would be helpful to have the PR description include the full scope.  Thanks.

File: server/server/server.go (hunk 3 above)
Comment (wiggin77) [reply server/server/server.go:236]: Ok, intentional but unrelated.

**Review by wiggin77 — APPROVED** [review]
LGTM :+1:   I request that the audit log field names be kept standard (don't add redundant fields), but everything else looks great.

File: server/api/api.go (hunk 2 above)
Comment (harshilsharma63) [reply server/api/api.go]: Okay, that makes sense. I've removed these meta fields from here now as they are already being included in `LogRecord`.

**Comment by harshilsharma63** [comment]
> LGTM 👍 I request that the audit log field names be kept standard (don't add redundant fields), but everything else looks great.

Done.
